File: Assets/Lua/game/dialog/Dialog.lua

---------------------------------------------------------------------------------------------------
--
--filename: game.dialog.Dialog
--date:2019/11/20 10:42:24
--author:heguang
--desc:弹窗管理
--
---------------------------------------------------------------------------------------------------
local strClassName = 'Dialog'
local Dialog = lua_declare(strClassName, lua_class(strClassName))
Dialog.currentAlert = nil
Dialog.alertItemDict = {}
Dialog.isPlayEffect = false
Dialog.playTimer = 0
Dialog.playTime = 0.8

function Dialog.Update()
	if Dialog.isPlayEffect then
		Dialog.playTimer = Dialog.playTimer + Time.deltaTime
		local t = Dialog.playTime
		if Dialog.currentAlert ~= nil then
			t = Dialog.currentAlert:GetTotalTime()
		end
		if (Dialog.playTimer > t) then
			Dialog.playTimer = 0
			Dialog.isPlayEffect = false

			if #Dialog.alertItemDict > 0 then
				if Dialog.currentAlert == nil then
					local msg =Dialog.alertItemDict[1]
					table.remove(Dialog.alertItemDict,1)
					local go = ObjectPool.Spawn("base/alerttext.u3d", "alerttext")
					UIManager.Instance:AddAlert(go)
					go.transform:SetAsFirstSibling()
					GameUtil.SetLayer(go.transform, LayerManager.UI)
					Dialog.currentAlert = LuaGameUtil.DoFile(go, "game.dialog.AlertController")
					Dialog.currentAlert:Show(msg.msg, msg.sp)
					Dialog.playTimer = 0
					Dialog.isPlayEffect = true
				end
			else
				Dialog.isPlayEffect = false
			end
		end
	end
end

function Dialog.Alert(message, spriteName)
	if #Dialog.alertItemDict > 10 then
		table.remove(Dialog.alertItemDict, 1)
	end
	--相同的提示就不弹了
	local res = false
	if spriteName == nil then
		if Dialog.currentAlert ~= nil then
			res = ( Dialog.currentAlert.message == message )
		end
		if not res then
			for _,v in ipairs(Dialog.alertItemDict) do
				if v.msg == message then
					res = true
					break
				end
			end
		end
	end
	if not res then
		table.insert(Dialog.alertItemDict, {msg=message, sp=spriteName})
		if not Dialog.isPlayEffect then
			Dialog.playTimer = Dialog.playTime
			Dialog.isPlayEffect = true
		end
	end
end

function Dialog.ClearAlert(go)
	Dialog.currentAlert = nil
	if not Dialog.isPlayEffect then
		Dialog.playTimer = Dialog.playTime
		Dialog.isPlayEffect = true
	end
end

function Dialog.ClearAllAlert()
	if Dialog.currentAlert ~= nil then
		Dialog.currentAlert:Remove()
	end
	Dialog.currentAlert = nil
	Dialog.alertItemDict = {}
end

function Dialog.ShowMessage(message,closeCallback, showOK, showCancel, showClose)
	if showOK == nil then showOK = true end
	if showCancel == nil then showCancel = false end
	if showClose == nil then showClose = true end
	UI.Dialog.ShowMessage(message,closeCallback, showOK, showCancel, showClose)
end

function Dialog.ShowWaiting(message)
	UI.Dialog.ShowWaiting(message)
end

function Dialog.CloseWaiting()
	UI.Dialog.CloseWaiting()
end

function Dialog.Show(go, isModal, closeCallback, showCallback)
	if isModal == nil then
		isModal = true
	end
	UI.Dialog.Show(go,isModal,closeCallback, showCallback)
end

function Dialog.Hide(state, callBack)
	if state == nil then
		state = DialogState.OK
	end
	UI.Dialog.Hide(state,callBack)
end

return Dialog

File: Assets/Script/Core/UI/Dialog.cs
using UnityEngine;
using System.Collections;
using UnityEngine.UI;
using System.Collections.Generic;

namespace UI
{
    [SLua.CustomLuaClass]
    public class Dialog
    {
        public delegate void DialogShowedDelegate(UIDialogView source);

        public delegate void DialogCloseDelegate(UIDialogView source, DialogState state);

        #region member

        private static UIDialogView mTargetDialog;
        private static UIDialogView mNextDialog;

        private static UIAlertView mCurrentAlert;

        private static GameObject mWaiting;

        #endregion

        #region public

        public static void Reset()
        {
            mTargetDialog = null;
            mNextDialog = null;
        }

        public static bool isShow
        {
            get
            {
                return Dialog.mTargetDialog != null;
            }
        }

        public static void Alert(string message)
        {
            if (mCurrentAlert != null)
            {
                mCurrentAlert.PlayStopAnim(delegate ()
                {
                    mCurrentAlert = null;
                    AlertEx(message);
                });
            }
            else
            {
                AlertEx(message);
            }
        }

        private static void AlertEx(string message)
        {
            GameObject go = ObjectPool.Spawn("base/alerttext.u3d", "alerttext");
            Text tfText = go.GetComponentInChildren<Text>();
            if (tfText != null)
            {
                tfText.text = message;
            }
            UIManager.Instance.AddAlert(go);
            mCurrentAlert = go.AddMissingComponent<UIAlertView>();
            mCurrentAlert.PlayAnim(RemoveAlert);
        }

        private static void RemoveAlert()
        {
            mCurrentAlert = null;
        }

        public static void ShowMessage(string message)
        {
            ShowMessage(message, null);
        }

        public static void ShowMessage(string message, Dialog.DialogCloseDelegate closeCallback)
        {
            ShowMessage(message, closeCallback, true, true, true);
        }

        public static void ShowMessage(string message, Dialog.DialogCloseDelegate closeCallback, bool showOK, bool showCancel, bool showClose)
        {
            GameObject go = ObjectPool.Spawn("base/messagepopup.u3d", "messagepopup");
            Transform tfText = go.transform.Find("Text");
            if (tfText != null)
            {
                Text text = tfText.GetComponent<Text>();
                if (text != null)
                {
                    text.text = message;
                }
            }

            Transform btnOK = go.transform.Find("btn/ButtonOK");
            if (btnOK != null)
            {
                btnOK.gameObject.SetActive(showOK);
            }

            Transform btnCancel = go.transform.Find("btn/ButtonCancel");
            if (btnCancel != null)
            {
                btnCancel.gameObject.SetActive(showCancel);
            }

            Transform btnClose = go.transform.Find("ButtonClose");
            if (btnClose != null)
            {
                btnClose.gameObject.SetActive(showClose);
            }

            Show(go, true, closeCallback);
        }

        public static void ShowWaiting()
        {
            Dialog.ShowWaiting(string.Empty);
        }

        public static void ShowWaiting(string message)
        {
            if(message == null)
            {
                message = string.Empty;
            }
            if (Dialog.mWaiting == null)
            {
                Dialog.mWaiting = ObjectPool.Spawn("base/waiting.u3d", "waiting");
                UIManager.Instance.AddWaiting(Dialog.mWaiting);
            }

            RectTransform tf = Dialog.mWaiting.transform as RectTransform;
            if (tf != null)
            {
                tf.offsetMax = Vector2.zero;
                tf.offsetMin = Vector2.zero;
                tf.anchorMin = Vector2.zero;
                tf.anchorMax = Vector2.one;
            }

            Transform tfText = Dialog.mWaiting.transform.Find("Text");
            if (tfText != null)
            {
                Text text = tfText.GetComponent<Text>();
                if (text != null)
                {
                    text.text = message;
                }
            }
        }

        public static void CloseWaiting()
        {
            if (Dialog.mWaiting != null)
            {
                ObjectPool.Recycle(Dialog.mWaiting);
                Dialog.mWaiting = null;
            }
        }

        public static void Show(GameObject go)
        {
            Show(go, true, null);
        }

        public static void Show(GameObject go, bool isModal, Dialog.DialogCloseDelegate closeCallback)
        {
            Show(go, isModal, closeCallback, null);
        }

        public static void Show(GameObject go, bool isModal, Dialog.DialogCloseDelegate closeCallback, Dialog.DialogShowedDelegate showCallback)
        {
            if (go == null)
            {
                return;
            }
            UIDialogView dialog = go.GetComponent<UIDialogView>();
            if (dialog == null)
            {
                dialog = go.AddComponent<UIDialogView>();
            }
            if (closeCallback != null)
            {
                dialog.hideCallback += closeCallback;
            }
            if(showCallback != null)
            {
                dialog.ShowedEvent += showCallback;
            }
            dialog.mIsModal = isModal;

            if (Dialog.mNextDialog != null)
            {
                Dialog.mNextDialog.Remove();
                Dialog.mNextDialog = null;
            }
            Dialog.mNextDialog = dialog;

            Hide(DialogState.OK, delegate
            {
                ShowEx();
            });
        }

        public static void Hide()
        {
            Hide(DialogState.OK, null);
        }

        public static void Hide(DialogCloseDelegate m)
        {
            Hide(DialogState.OK, m);
        }

        public static void Hide(DialogState state, DialogCloseDelegate m = null)
        {
            if (Dialog.mTargetDialog == null)
            {
                if (m != null)
                {
                    m(null, state);
                }
                return;
            }
            if (m != null)
            {
                Dialog.mTargetDialog.hideCallback += m;
            }
            Dialog.mTargetDialog.Hide(state);
        }

        #endregion

        #region private

        private static void ShowEx()
        {
            if (Dialog.mNextDialog == null)
            {
                return;
            }
            Dialog.mTargetDialog = Dialog.mNextDialog;
            Dialog.mNextDialog = null;

            if (Dialog.mTargetDialog.mIsModal)
            {
                UIManager.Instance.ShowModal(true);
            }
            else
            {
                UIManager.Instance.HideModal(true);
            }
            UIManager.Instance.AddDialog(Dialog.mTargetDialog.gameObject);

            Dialog.mTargetDialog.HideEvent += DialogHide;
            Dialog.mTargetDialog.Show();
        }

        private static void DialogHide(UIDialogView dialog, DialogState state)
        {
            UIManager.Instance.HideModal(true);
            Dialog.mTargetDialog = null;
        }

        #endregion
    }
}

[SLua.CustomLuaClass]
public enum DialogState
{
    OK,
    CANCEL,
    CLOSE
}
